### PR TITLE
feat(router): add an option to rerun guards and resolvers when query …

### LINKED ISSUE #14514: Rerun guards when query params change
**I'm submitting a ...**  (check one with "x")
```
[ ] bug report => search github for a similar issue or PR before submitting
[x] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

## Problems

Currently a route will be activated (and as a result its guards and resolvers will run) if and only if its matrix parameters or positional parameters change. This is a reasonable default, but there are two other scenarios that can be common:

1. Route can depend on query params and fragment. And, as a result, we would like rerun its guards and resolvers when query params change. This is common for routers at the top of the router configuration.

2. Route can depend on some transient client state not captured in the URL. As, as a result, we would like to rerun its guards and resolvers any time any navigation happens.

### Design 1: Fix set of options

```
{path: 'root', runGuardsAndResolvers: MATRIX_AND_QUERY_PARAMS_CHANGE, component: A, resolvers: {my: myResolver}}
```

And have the union type defined like this: 

```
type RunGuardsAndResolvers = MATRIX_PARAMS_CHANGE | MATRIX_AND_QUERY_PARAMS_CHANGE | ALWAYS;
```

### Design 2: Pass functions

```
{path: 'root', runGuardsAndResolvers: matrixAndQueryParamsChange, component: A, resolvers: {my: myResolver}}
```
```
type ActivateWhenFn = (old: RouterStateSnaphost, new: RouterStateSnaphost) => boolean;
```

### PR DESCRIPTION
**Pull request by vsavkin**
…changes

Closes #14514

**What kind of change does this PR introduce?** (check one with "x")
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)

See issue #14514 for more info.

**What is the new behavior?**

**Does this PR introduce a breaking change?** (check one with "x")
```
[ ] Yes
[ ] No
```

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

File: modules/@angular/router/src/config.ts [hunk 1]
```diff
@@ -35,6 +38,10 @@ import {UrlSegment, UrlSegmentGroup} from './url_tree';
  * - `data` is additional data provided to the component via `ActivatedRoute`.
  * - `resolve` is a map of DI tokens used to look up data resolvers. See {@link Resolve} for more
  *   info.
+ * - `runGuardsAndResolvers` defines when guards and resovlers will be run. By default they run only
+ *    when the matrix parameters of the route change. When set to `paramsOrQueryParamsChange` they
+ * will also
```
Comment (vicb) [inline modules/@angular/router/src/config.ts]: nit: fix formatting

File: modules/@angular/router/src/router.ts [hunk 2]
```diff
@@ -942,6 +942,25 @@ export class PreActivation {
...
+
+      case 'paramsChange':
+        return !equalParamsAndUrlSegments(curr, future);
+
+      case 'paramsOrQueryParamsChange':
+        return !equalParamsAndUrlSegments(curr, future) ||
+            !shallowEqual(curr.queryParams, future.queryParams);
+
+      default:
+        return !equalParamsAndUrlSegments(curr, future);
```
Comment (vicb) [inline modules/@angular/router/src/router.ts]: coalesce with `case 'paramsChange':`

File: modules/@angular/router/test/integration.spec.ts [hunk 3]
```diff
@@ -1527,6 +1527,125 @@ describe('Integration', () => {
...
+              provide: 'loggingCanActivate',
+              useValue: (a: any, b: any) => {
+                count++;
+                return true;
+              }
+            }]
+          });
+        });
+
+        it('should rererun guards when params change',
```
Comment (vicb) [inline modules/@angular/router/test/integration.spec.ts]: why twice :) (`rerun` should be enough)

Comment (vicb) [inline modules/@angular/router/test/integration.spec.ts]: matrix

**Comment by vicb** [comment]
LGTM with a few nits.

Thanks for the PR !

**Comment by vsavkin** [comment]
Thanks for reviewing it. Fixed them.

**Review by vicb — APPROVED** [review]

**Pull request closed without merging**